### PR TITLE
WIP: Build Mode

### PR DESCRIPTION
**Pull request by Zintixx**
This is very WIP but gets the basic infrastructure of building down.
Parsing ``UGCMap`` for all buyable land in the game
Parsing ``FurnishingShop`` for all buyable items when building
Adjusted parsing for `Map` to include saleable cubes.
EDIT: Added beginning stages of the Home button functionality

With the home button currently not working, the only build-able land right now would be land you can purchase in the world.

The biggest things that need to happen is:
- Create a `FurnishingInventory`. This inventory is all the blocks that you have placed in the world (interior and exterior of your home).
- Create a `WarehouseInventory`. This inventory is where all your blocks are stored that are _NOT_ placed anywhere.
- We need to have all blocks be redirected to the `WarehouseInventory` rather than in the standard player's inventory
- Some kind of system to have a specific map ID load all applicable cubes and status of purchased land.
#73

### CODE REVIEW

File: Maple2Storage/Types/Coord.cs [hunk 1]
```diff
@@ -213,16 +213,18 @@ public static CoordB From(sbyte x, sbyte y, sbyte z)
 
         public readonly CoordF ToFloat()
         {
-            return CoordF.From(X, Y, Z);
+            return CoordF.From(
+                (float) (X * Block.BLOCK_SIZE),
+                (float) (Y * Block.BLOCK_SIZE),
+                (float) (Z * Block.BLOCK_SIZE));
```
Comment (AlanMorel) [inline Maple2Storage/Types/Coord.cs:219]: Will this break compatibility with existing usages?

File: MapleServer2/PacketHandlers/Game/FishingHandler.cs [hunk 2]
```diff
@@ -111,17 +111,14 @@ private static void HandlePrepareFishing(GameSession session, PacketReader packe
             // Adding GuideObject
             CoordF guideBlock = GetObjectBlock(fishingBlocks, session.FieldPlayer.Coord);
             guideBlock.Z += Block.BLOCK_SIZE; // sits on top of the block
-            GuideObject guide = new GuideObject()
-            {
-                Type = 1,
-            };
+            GuideObject guide = new GuideObject(1, session.Player.CharacterId) { };
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/FishingHandler.cs:114]: Can get rid of the parenthesis at the end I think

File: MapleServer2/Packets/ResponseCubePacket.cs [hunk 3]
```diff
@@ -10,14 +11,111 @@ public static class ResponseCubePacket
...
+            pWriter.WriteEnum(ResponseCubePacketMode.CompletePurchase);
+            return pWriter;
+        }
+
+        public static Packet PlaceFurnishing(IFieldObject<Cube> cube)
         {
             PacketWriter pWriter = PacketWriter.Of(SendOp.RESPONSE_CUBE);
+            pWriter.WriteEnum(ResponseCubePacketMode.PlaceFurnishing);
+            pWriter.WriteBool(false); // unk bool
+            pWriter.WriteInt(cube.ObjectId); // objectId PALYER!
```
Comment (AlanMorel) [inline MapleServer2/Packets/ResponseCubePacket.cs:65]: PALYER!

File: MapleServer2/Packets/WarehouseInventoryPacket.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        private enum WarehouseInventoryPacketMode : byte
+        {
+            StartList = 0x1,
+            Count = 0x2,
+            Load = 0x3,
+            Add = 0x4,
+            UpdateQuantity = 0x5,
+            EndList = 0x8,
+        }
+        public static Packet StartList()
```
Comment (AlanMorel) [inline MapleServer2/Packets/WarehouseInventoryPacket.cs:19]: Add empty line above

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Packets/ResponseCubePacket.cs (hunk 3 above)
Comment (Zintixx) [reply MapleServer2/Packets/ResponseCubePacket.cs:65]: lmao left my notes here oops

File: Maple2Storage/Types/Coord.cs (hunk 1 above)
Comment (Zintixx) [reply Maple2Storage/Types/Coord.cs:219]: It shouldn't because nothing else was using it.

**Review by AlanMorel — APPROVED** [review]

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/RequestCubeHandler.cs [hunk 5]
```diff
@@ -18,8 +19,16 @@ public class RequestCubeHandler : GamePacketHandler
 
         private enum RequestCubeMode : byte
```
[the remark this answers is not in the document]
Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/RequestCubeHandler.cs:20]: It does seem to be a mishmash of different content. I'm not really sure what better name there could be for an opcode that handles both player housing plots and interacting with the world models.

File: MapleServer2/PacketHandlers/Game/RequestCubeHandler.cs [hunk 6]
```diff
@@ -28,13 +37,198 @@ public override void Handle(GameSession session, PacketReader packet)
...
+        private static bool HandlePlotPayment(GameSession session, int priceItemCode, int price)
+        {
+            switch (priceItemCode)
+            {
+                case 0:
+                    return true;
+                case 90000001:
+                case 90000002:
+                case 90000003:
+                    return !session.Player.Wallet.Meso.Modify(-price);
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/RequestCubeHandler.cs:118]:
```suggestion
                    return session.Player.Wallet.Meso.Modify(-price);
```

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

**Review by AngeloTadeucci — APPROVED** [review]

**Review by AlanMorel — APPROVED** [review]

**Pull request merged**